vnext/Chakra/ChakraJsiRuntime.cpp

```diff
@@ -0,0 +1,862 @@
+  evaluateJavaScriptSimple(*buffer, sourceURL);
+#endif
+}
+
+jsi::Object ChakraJsiRuntime::global() {
+  JsValueRef value;
+  JsGetGlobalObject(&value);
+  return createObject(value);
+}
+
+std::string ChakraJsiRuntime::description() {
+  if (desc_.empty()) {
+    desc_ = std::string("<ChakraJsiRuntime>");
+  }
+  return desc_;
+}
+
+bool ChakraJsiRuntime::isInspectable() {
+  return false;
+}
+
+ChakraJsiRuntime::ChakraPropertyIdValue::~ChakraPropertyIdValue() {
+  JsRelease(propId_, nullptr);
+}
+
+void ChakraJsiRuntime::ChakraPropertyIdValue::invalidate() {
+  delete this;
+}
+
+ChakraJsiRuntime::ChakraPropertyIdValue::ChakraPropertyIdValue(JsPropertyIdRef propIdRef)
+  : propId_(propIdRef) {
+  JsAddRef(propIdRef, nullptr);
+}
+
+ChakraJsiRuntime::ChakraStringValue::ChakraStringValue(JsValueRef str)
+  : str_(str) {
+  JsAddRef(str, nullptr);
+}
+
+void ChakraJsiRuntime::ChakraStringValue::invalidate() {
+  delete this;
+}
+
+ChakraJsiRuntime::ChakraStringValue::~ChakraStringValue() {
+  JsRelease(str_, nullptr);
+}
+
+ChakraJsiRuntime::ChakraObjectValue::ChakraObjectValue(
+  JsValueRef obj
+) : obj_(obj)
+{
+  JsAddRef(obj_, nullptr);
+}
+
+void ChakraJsiRuntime::ChakraObjectValue::invalidate() {
+  delete this;
+}
+
+ChakraJsiRuntime::ChakraObjectValue::~ChakraObjectValue() {
+  JsRelease(obj_, nullptr);
+}
+
+jsi::Runtime::PointerValue* ChakraJsiRuntime::cloneString(
+  const jsi::Runtime::PointerValue* pv) {
+
+  if (!pv) {
+    return nullptr;
+  }
+  const ChakraStringValue* string = static_cast<const ChakraStringValue*>(pv);
+  return makeStringValue(string->str_);
+}
+
+jsi::Runtime::PointerValue* ChakraJsiRuntime::cloneObject(
+  const jsi::Runtime::PointerValue* pv) {
+
+  if (!pv) {
+    return nullptr;
+  }
+
+  const ChakraObjectValue* object = static_cast<const ChakraObjectValue*>(pv);
+  return makeObjectValue(object->obj_);
+}
+
+jsi::Runtime::PointerValue* ChakraJsiRuntime::clonePropNameID(
+  const jsi::Runtime::PointerValue* pv) {
+
+  if (!pv) {
+    return nullptr;
+  }
+
+  const ChakraPropertyIdValue* propId = static_cast<const ChakraPropertyIdValue*>(pv);
+  return makePropertyIdValue(propId->propId_);
+}
+
+jsi::PropNameID ChakraJsiRuntime::createPropNameIDFromAscii(
+  const char* str,
+  size_t length) {
+
+  JsValueRef propIdRef = CreateJSPropertyId(str, length);
+
+  auto res = createPropNameID(propIdRef);
+  return res;
+}
+
+jsi::PropNameID ChakraJsiRuntime::createPropNameIDFromUtf8(
+  const uint8_t* utf8,
+  size_t length) {
+
+  JsValueRef prpoIdRef = CreateJSPropertyId(reinterpret_cast<const char*>(utf8), length);
+  auto res = createPropNameID(prpoIdRef);
+  return res;
+}
+
+jsi::PropNameID ChakraJsiRuntime::createPropNameIDFromString(const jsi::String& str) {
+  std::string propNameString = JSStringToSTLString(stringRef(str));
+  return createPropNameIDFromUtf8(reinterpret_cast<const uint8_t*>(propNameString.c_str()), propNameString.length());

```

Comment:
>createPropNameIDFromUtf8 [](start = 9, length = 24)

I guess this is chakra api but odd that creating a prop name takes a utf8 string, but getting a property name returns utf16.

Reply:
ChakraCore.h usually has extra methods which takes 8-bit string .. it's there for string and propId creation .. But so far, I can't see one for propId to name .. asymmetry ..